app/Http/Controllers/Account/GithubUsersController.php

```diff
@@ -0,0 +1,125 @@
+        $this->middleware('verify-user');
+
+        parent::__construct();
+    }
+
+    public function callback()
+    {
+        $params = get_params(request()->all(), null, [
+            'code:string',
+            'error:string',
+            'state:string',
+        ], ['null_missing' => true]);
+
+        abort_if($params['state'] === null, 422, 'Missing state parameter.');
+        abort_unless(
+            hash_equals(session()->pull('github_auth_state', ''), $params['state']),
+            403,
+            'Invalid state.',
+        );
+
+        // If the user denied authorization on GitHub, redirect back to the GitHub account settings
+        // <https://docs.github.com/en/apps/oauth-apps/maintaining-oauth-apps/troubleshooting-authorization-request-errors#access-denied>
+        if ($params['error'] === 'access_denied') {
+            return redirect(route('account.edit').'#github');
+        }
+
+        abort_if($params['error'] !== null, 500, 'Error obtaining authorization from GitHub.');
+        abort_if($params['code'] === null, 422, 'Missing code parameter.');
+
+        try {
+            $token = $this
+                ->makeGithubOAuthProvider()
+                ->getAccessToken('authorization_code', ['code' => $params['code']]);
+        } catch (GithubIdentityProviderException $exception) {
+            // <https://docs.github.com/en/apps/oauth-apps/maintaining-oauth-apps/troubleshooting-oauth-app-access-token-request-errors#bad-verification-code>
+            abort_if(
+                $exception->getMessage() === 'bad_verification_code',
+                422,
+                'Invalid authorization code.',
+            );
+
+            // <https://docs.github.com/en/apps/oauth-apps/maintaining-oauth-apps/troubleshooting-oauth-app-access-token-request-errors#unverified-user-email>
+            abort_if(
+                $exception->getMessage() === 'unverified_user_email',
+                422,
+                osu_trans('accounts.github_user.error.unverified_email'),
+            );
+
+            throw $exception;
+        }
+
+        $client = new \Github\Client();
+        $client->authenticate($token->getToken(), \Github\AuthMethod::ACCESS_TOKEN);
+        $apiUser = $client->currentUser()->show();
+
+        $user = GithubUser::firstWhere('canonical_id', $apiUser['id']);
+
+        abort_if($user === null, 422, osu_trans('accounts.github_user.error.no_contribution'));
+        abort_if($user->user_id !== null, 422, osu_trans('accounts.github_user.error.already_linked'));
+
+        $user->update([
+            'user_id' => auth()->id(),
+            'username' => $apiUser['login'],
+        ]);
+
+        return redirect(route('account.edit').'#github');
+    }
+
+    public function create()
+    {
+        abort_unless(GithubUser::canAuthenticate(), 404);
+        abort_if(
+            auth()->user()->githubUser()->exists(),
+            422,
+            'Cannot link more than one GitHub account.',
```

Comment:
for this one, I changed it so that you get redirected back to the account page (where you should see your existing github account if you didn't already for whatever reason)

for the other untranslated error messages, I left them like that because they should never be seen if not purposely messing with the request from github... seems kinda pointless to translate?